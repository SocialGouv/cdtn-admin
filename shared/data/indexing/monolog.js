import { Client } from "@elastic/elasticsearch";
import { Queries } from "@socialgouv/cdtn-monolog";
import { getRouteBySource, SOURCES } from "@socialgouv/cdtn-sources";

// import { logger } from "./logger";

const ES_LOGS = process.env.ES_LOGS;
const ES_LOGS_TOKEN = process.env.ES_LOGS_TOKEN;

if (!ES_LOGS || !ES_LOGS_TOKEN) {
  if (process.env.NODE_ENV != "test") {
    console.error(
      `Missing env variable for accessing Monolog Elastic Search logs : ${
        ES_LOGS ? "" : "ES_LOGS"
      } ${
        ES_LOGS_TOKEN ? "" : "ES_LOGS_TOKEN"
      }, Covisites won't be available in related items.`
      // logger.error(
      //   `Missing env variable for accessing Monolog Elastic Search logs : ${
      //     ES_LOGS ? "" : "ES_LOGS"
      //   } ${ES_LOGS_TOKEN ? "" : "ES_LOGS_TOKEN"}`
      // );
    );
  }
} else {
  console.error(`Accessing Monolog Elastic Search logs on ${ES_LOGS}`);
}

const esClientConfig = {
  auth: { apiKey: ES_LOGS_TOKEN },
  node: ES_LOGS,
};

const client =
  ES_LOGS && ES_LOGS_TOKEN ? new Client(esClientConfig) : undefined;

const queries = Queries(client, "log_reports");

export const fetchCovisits = async (doc) => {
  let sourceRoute = getRouteBySource(doc.source);

  // special case for fiches MT
  if (doc.source == SOURCES.SHEET_MT_PAGE) {
    sourceRoute = getRouteBySource(SOURCES.SHEET_MT);
  }

  const path = `${sourceRoute}/${doc.slug}`;
  const links = await queries
    .getCovisitLinks(path)
    .then((covisits) => covisits.links)
    .catch(() => {
      // TODO avoid silent and deal with failure properly
      // console.error(err);
      return undefined;
    });

  doc.covisits = links;

  return doc;
};
