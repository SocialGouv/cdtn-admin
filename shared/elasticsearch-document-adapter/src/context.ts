//@ts-expect-error
import nctx from "nctx";

export const context = nctx.create(Symbol("app"));
