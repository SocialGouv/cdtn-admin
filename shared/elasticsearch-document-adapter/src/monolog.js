import { Client } from "@elastic/elasticsearch";
import { logger } from "@socialgouv/cdtn-logger";
import { getRouteBySource, SOURCES } from "@socialgouv/cdtn-sources";

const COVISIT_BATCH_SIZE = 500;

const reportType = "covisit";

const covisitsMap = new Map();

export const buildCovisitMap = async (esLogs, esLogsToken) => {
  const esClientConfig = {
    auth: { apiKey: esLogsToken },
    node: esLogs,
  };

  const client = new Client(esClientConfig);

  const query = { term: { reportType } };

  const count = (
    await client.count({
      body: { query },
      index: "log_reports",
    })
  ).body.count;

  for (let b = 0; b <= Math.floor(count / COVISIT_BATCH_SIZE); b++) {
    const results = await client.search({
      body: { query },
      from: b * COVISIT_BATCH_SIZE,
      index: "log_reports",
      size: COVISIT_BATCH_SIZE,
    });
    const hits = results.body.hits?.hits;
    hits.forEach(({ _source: { content, links } }) => {
      covisitsMap.set(content, links);
    });
  }

  logger.info(`${covisitsMap.size} covisits found`);
};

export const addCovisits = (doc) => {
  let sourceRoute = getRouteBySource(doc.source);

  // special case for fiches MT
  if (doc.source == SOURCES.SHEET_MT_PAGE) {
    sourceRoute = getRouteBySource(SOURCES.SHEET_MT);
  }

  const path = `${sourceRoute}/${doc.slug}`;

  doc.covisits = covisitsMap.get(path);
};
