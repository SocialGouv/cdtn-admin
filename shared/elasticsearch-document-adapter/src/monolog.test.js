import { SOURCES } from "@socialgouv/cdtn-sources";

import { context } from "./context";
import { fetchCovisits } from "./monolog";

const testDoc = {
  slug: "activite-partielle-chomage-partiel",
  source: SOURCES.SHEET_MT,
};

jest.mock("@socialgouv/cdtn-monolog", () => ({
  LogQueries: () => ({
    getCovisitLinks: async (path) => {
      if (
        path == "fiche-ministere-travail/activite-partielle-chomage-partiel"
      ) {
        return Promise.resolve({ links: [] });
      } else {
        return Promise.reject();
      }
    },
  }),
}));

describe("Test covisits are added if available.", () => {
  test("should add covisites to item", async () => {
    context.provide();
    const res = await fetchCovisits(testDoc);
    expect(res).toMatchSnapshot();
  });

  test("should not fail with other item", async () => {
    context.provide();
    const res = await fetchCovisits({ slug: "fake", source: SOURCES.SHEET_MT });
    expect(res).toMatchSnapshot();
  });
});
