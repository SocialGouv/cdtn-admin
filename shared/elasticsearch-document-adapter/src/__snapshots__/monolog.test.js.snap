// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test covisits are added if available. should add covisites to item 1`] = `
Object {
  "covisits": Array [],
  "slug": "activite-partielle-chomage-partiel",
  "source": "fiches_ministere_travail",
}
`;

exports[`Test covisits are added if available. should not fail with other item 1`] = `
Object {
  "covisits": undefined,
  "slug": "fake",
  "source": "fiches_ministere_travail",
}
`;
